### PR TITLE
Implement Pagination

### PR DESCRIPTION
**Pull request by AllenFang**
#135

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,13 +56,13 @@
   "dependencies": {
     "classnames": "^2.2.5",
     "prop-types": "^15.5.10",
-    "react": "^15.6.1",
-    "react-dom": "^15.6.1"
+    "react": "^16.0.0",
```
Comment (AllenFang) [inline package.json]: Why: https://til.hashrocket.com/posts/oqjntperif-render-an-array-of-elements-with-react-16

File: package.json [hunk 2]
```diff
@@ -56,13 +56,13 @@
...
+    "react-dom": "^16.0.0"
   },
   "peerDependencies": {
     "prop-types": "^15.0.0",
-    "react": "^15.0.0",
-    "react-dom": "^15.0.0"
+    "react": "^15.0.0 || ^16",
+    "react-dom": "^15.0.0 || ^16"
```
Comment (AllenFang) [inline package.json]: I'll consider to make `^16` only as peer when we release, but I need some test

File: packages/react-bootstrap-table2-example/examples/pagination/index.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  text: 'Product ID'
+}, {
+  dataField: 'name',
+  text: 'Product Name'
+}, {
+  dataField: 'price',
+  text: 'Product Price'
+}];
+
+<BootstrapTable keyField='id' data={ products } columns={ columns } />
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-example/examples/pagination/index.js]: forget to update code

File: packages/react-bootstrap-table2-example/examples/pagination/pagination-hooks.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  text: 'Product ID'
+}, {
+  dataField: 'name',
+  text: 'Product Name'
+}, {
+  dataField: 'price',
+  text: 'Product Price'
+}];
+
+<BootstrapTable keyField='id' data={ products } columns={ columns } />
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-example/examples/pagination/pagination-hooks.js]: forgot to update code

File: packages/react-bootstrap-table2-paginator/src/page-resolver.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
+/* eslint no-mixed-operators: 0 */
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-paginator/src/page-resolver.js]: this class is main class for calculate page information by `props`

File: packages/react-bootstrap-table2-paginator/src/pagination.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    this.toggleDropDown = this.toggleDropDown.bind(this);
+    this.handleChangePage = this.handleChangePage.bind(this);
+    this.handleChangeSizePerPage = this.handleChangeSizePerPage.bind(this);
+    this.state = this.initialState();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { dataSize, currSizePerPage } = nextProps;
+
+    if (currSizePerPage !== this.props.currSizePerPage || dataSize !== this.props.dataSize) {
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-paginator/src/pagination.js]: `currSizePerPage` and `dataSize` will effect on `totalPages` and `lastPage`

File: packages/react-bootstrap-table2-paginator/src/wrapper.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    render() {
+      const { pagination: { Pagination }, options, store } = this.props;
+      const { currPage, currSizePerPage } = this.state;
+      const base = baseElement({
+        ...this.props,
+        key: 'table',
+        data: store.getByCurrPage(currPage, currSizePerPage)
+      });
+
+      return [
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-paginator/src/wrapper.js]: there's a kind of different than previous wrapper, 
1. In this class, I export a factory function for creating pagination wrapper
2. This factory function will have two args: `baseElement` and `Const` which are passed from `react-bootstrap-table2`(`table-factory`)

3. This HOC component is not render next element, instead, I render next element and Pagination component together

File: packages/react-bootstrap-table2-paginator/src/wrapper.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      });
+    }
+
+    render() {
+      const { pagination: { Pagination }, options, store } = this.props;
+      const { currPage, currSizePerPage } = this.state;
+      const base = baseElement({
+        ...this.props,
+        key: 'table',
+        data: store.getByCurrPage(currPage, currSizePerPage)
```
Comment (AllenFang) [inline packages/react-bootstrap-table2-paginator/src/wrapper.js]: slice data by page here

File: packages/react-bootstrap-table2/src/container.js [hunk 9]
```diff
@@ -64,6 +65,8 @@ const withDataStore = Base =>
         return wrapWithSelection(baseProps);
       } else if (this.props.columns.filter(col => col.sort).length > 0) {
         return wrapWithSort(baseProps);
+      } else if (this.props.pagination) {
+        return wrapWithPagination(baseProps);
```
Comment (AllenFang) [inline packages/react-bootstrap-table2/src/container.js]: use wrapper as usual

File: packages/react-bootstrap-table2/src/table-factory.js [hunk 10]
```diff
@@ -19,8 +20,18 @@ export const wrapWithSort = props =>
 export const pureTable = props =>
   React.createElement(BootstrapTable, { ...props });
 
+export const wrapWithPagination = (props) => {
+  if (props.pagination) {
+    const { wrapper } = props.pagination;
+    const PaginationBase = wrapper(pureTable, Const);
+    return React.createElement(PaginationBase, { ...props });
```
Comment (AllenFang) [inline packages/react-bootstrap-table2/src/table-factory.js]: 1. get pagination `wrapper` from props
2. call the wrapper factory function (`wrapper` is factory function`)

File: packages/react-bootstrap-table2-example/examples/pagination/index.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  text: 'Product Name'
+}, {
+  dataField: 'price',
+  text: 'Product Price'
+}];
+
+<BootstrapTable keyField='id' data={ products } columns={ columns } pagination={ paginator } />
+`;
+
+class BasicPaginationTable extends React.PureComponent {
```
Comment (Chun-MingChen) [inline packages/react-bootstrap-table2-example/examples/pagination/index.js]: I just wondered what kind of difficulties you faced and you selected `PureComponent ` here :)?

Comment (AllenFang) [reply packages/react-bootstrap-table2-example/examples/pagination/index.js]: Nothing there, I just test some state change manually but forgot to remove it, thanks this reminder

**Comment by AllenFang** [comment]
umm.... huge PR, review by commit will be little better lol.

**Comment by AllenFang** [comment]
@Chun-MingChen after merged, you are suppose to run `yarn install` again on your local.

**Pull request merged**